README.md
# Mule Anypoint Connector for Couchbase Database

This connector allows Mule ESB to connect to [Couchbase](https://www.couchbase.com/) NoSQL database and perform CRUD operations on Couchbase database.

[![CircleCI](https://circleci.com/gh/manikmagar/mule-couchbase-connector/tree/dev.svg?style=svg)](https://circleci.com/gh/manikmagar/mule-couchbase-connector/tree/dev)

# Mule supported versions
Mule 3.5+


# Couchbase supported versions
[Couchbase Server 3+](https://www.couchbase.com/downloads).

## Couchbase Services supported
* Couchbase Key-Value Service

## Couchbase Libraries
* Couchbase Java Client - 2.4.5

# Installation
For beta connectors you can download the source code and build it with devkit to find it available on your local repository. Then you can add it to Studio

For released connectors you can download them from the update site in Anypoint Studio.
Open Anypoint Studio, go to Help → Install New Software and select Anypoint Connectors Update Site where you’ll find all available connectors.

#Usage
For information about usage our documentation at http://github.com/manikmagar/mule-couchbase-connector.

# Reporting Issues

We use GitHub:Issues for tracking issues with this connector. You can report new issues at this link http://github.com/manikmagar/mule-couchbase-connector/issues.


CHANGELOG.md
# Couchbase Release Notes
# --------------------------------------------------------------------
# Date: <DD-MMM-YYYY>
Example: 24-Jan-2015
# Version: (default value: 0.1)
# Supported API versions: 
Example: Salesforce Bulk API v.28, SOAP API V.28 (include links to API docs)
# Supported Mule Runtime Versions: 
${project.devkitVersion}
# New Features and Functionality
(default value: Initial version) 
Example: Added support for Query. http://github.com/mulesoft/couchbase/issues/61
# Closed Issues in this release 
(include issue link when possible)
Example: Fixed problem with lost telefone on Bulk Update (CLDCONNECT-123).
# Known Issues in this release
Example: Session ID expires after  30 days. http://github.com/mulesoft/couchbase/issues/62


src/test/java/org/mule/modules/couchbase/automation/runner/FunctionalTestSuite.java
/**
 * The software in this package is published under the terms of the Apache v2.0 license,
 * a copy of which has been included with this distribution in the LICENSE.md file.
 */
package org.mule.modules.couchbase.automation.runner;

import java.util.ArrayList;

import javax.validation.constraints.NotNull;

import org.apache.http.HttpResponse;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.utils.URIBuilder;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.junit.runners.Suite.SuiteClasses;
import org.mule.modules.couchbase.CouchbaseConnector;
import org.mule.modules.couchbase.automation.functional.DeleteDocumentTestCases;
import org.mule.modules.couchbase.automation.functional.ExecuteQueryTestCase;
import org.mule.modules.couchbase.automation.functional.GetDocumentTestCases;
import org.mule.modules.couchbase.automation.functional.InsertDocumentTestCases;
import org.mule.modules.couchbase.automation.functional.UnlockDocumentTestCases;
import org.mule.modules.couchbase.automation.functional.UpdateDocumentTestCases;
import org.mule.modules.couchbase.automation.functional.UpsertDocumentTestCases;
import org.mule.modules.couchbase.config.CouchbaseClusterClient;
import org.mule.modules.couchbase.config.CouchbaseConnectorConfig;
import org.mule.tools.devkit.ctf.mockup.ConnectorTestContext;

import com.couchbase.client.java.Cluster;
import com.couchbase.mock.Bucket.BucketType;
import com.couchbase.mock.BucketConfiguration;
import com.couchbase.mock.CouchbaseMock;
import com.couchbase.mock.JsonUtils;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

@RunWith(Suite.class)
@SuiteClasses({

GetDocumentTestCases.class,
UpdateDocumentTestCases.class,
InsertDocumentTestCases.class,
UpsertDocumentTestCases.class,
DeleteDocumentTestCases.class,
UnlockDocumentTestCases.class,
ExecuteQueryTestCase.class
})

public class FunctionalTestSuite {

	public static CouchbaseMock couchbaseMock;
	public static com.couchbase.client.java.Bucket bucket;

	protected static final BucketConfiguration bucketConfiguration = new BucketConfiguration();

	protected static Cluster cluster;

	protected static int carrierPort;
	protected static int httpPort;
	protected static CouchbaseClusterClient client;

	protected static void getPortInfo(String bucket) throws Exception {
		httpPort = couchbaseMock.getHttpPort();
		URIBuilder builder = new URIBuilder();
		builder.setScheme("http").setHost("localhost").setPort(httpPort).setPath("mock/get_mcports")
				.setParameter("bucket", bucket);
		HttpGet request = new HttpGet(builder.build());
		HttpClient client = HttpClientBuilder.create().build();
		HttpResponse response = client.execute(request);
		int status = response.getStatusLine().getStatusCode();
		if (status < 200 || status > 300) {
			throw new ClientProtocolException("Unexpected response status: " + status);
		}
		String rawBody = EntityUtils.toString(response.getEntity());
		JsonObject respObject = JsonUtils.GSON.fromJson(rawBody, JsonObject.class);
		JsonArray portsArray = respObject.getAsJsonArray("payload");
		carrierPort = portsArray.get(0).getAsInt();
	}

	protected static void createMock(@NotNull String name, @NotNull String password) throws Exception {
		bucketConfiguration.numNodes = 1;
		bucketConfiguration.numReplicas = 1;
		bucketConfiguration.numVBuckets = 1024;
		bucketConfiguration.name = name;
		bucketConfiguration.type = BucketType.COUCHBASE;
		bucketConfiguration.password = password;
		ArrayList<BucketConfiguration> configList = new ArrayList<BucketConfiguration>();
		configList.add(bucketConfiguration);
		couchbaseMock = new CouchbaseMock(0, configList);
		couchbaseMock.start();
		couchbaseMock.waitForStartup();

	}

	protected static void createClient() {
		client = CouchbaseClusterClient.get();
		CouchbaseConnectorConfig config = new CouchbaseConnectorConfig();
		config.setBootstrapHttpDirectPort(httpPort);
		config.setBootstrapCarrierDirectPort(carrierPort);
		cluster = client.createCluster("127.0.0.1", config);
		bucket = cluster.openBucket("default");
	}


	@BeforeClass
	public static void initialiseSuite() throws Exception{

		createMock("default", "");
		getPortInfo("default");
		createClient();
		ConnectorTestContext.initialize(CouchbaseConnector.class);
	}

	@AfterClass
    public static void shutdownSuite() {
		if (cluster != null) {
			System.out.println("Stop cluster");
			bucket.close();
			cluster.disconnect();
		}
		if(client.isConnected()){
			System.out.println("Disconnect client");
			client.disconnectCluster();
		}
		if (couchbaseMock != null) {
			System.out.println("Stop mock");
			couchbaseMock.stop();
		}
    	ConnectorTestContext.shutDown();
    }


}


src/test/java/org/mule/modules/couchbase/automation/functional/ExecuteQueryTestCase.java
package org.mule.modules.couchbase.automation.functional;

import org.junit.Test;
import org.mule.modules.couchbase.automation.runner.CouchbaseAbstractTestCase;

public class ExecuteQueryTestCase extends CouchbaseAbstractTestCase {

	
	@Test
	public void testExecuteQuery(){
		
		//CouchbaseMock does not support N1QL queries.
		
//		String query = "select * from default where type = 'user'";
//		org.mule.api.MuleEvent muleEvent = null;
//		
//		List<Map<String, Object>> lst = getConnector().executeQuery(muleEvent, query, Collections.EMPTY_LIST);
//		
//		assertNotNull(lst);
//		
	}
	
}


src/test/java/org/mule/modules/couchbase/automation/functional/GetDocumentTestCases.java
/**
 * The software in this package is published under the terms of the Apache v2.0 license,
 * a copy of which has been included with this distribution in the LICENSE.md file.
 */
package org.mule.modules.couchbase.automation.functional;

import org.hamcrest.MatcherAssert;
import org.hamcrest.Matchers;
import org.junit.Test;
import org.mule.modules.couchbase.automation.runner.CouchbaseAbstractTestCase;
import org.mule.modules.couchbase.model.JavaMapDocument;

import com.couchbase.client.java.error.CASMismatchException;

public class GetDocumentTestCases extends CouchbaseAbstractTestCase {


	@Test
	public void verifyNoDocumentReturn() {
		org.mule.api.MuleEvent muleEvent = null;
		java.lang.String id = "1";
		boolean lockDocument = false;
		int lockDuration = 0;
		boolean refreshExpirationTime = false;
		int refreshTime = 60;
		JavaMapDocument returnDoc = getConnector().getDocument(muleEvent, id, lockDocument, lockDuration, refreshExpirationTime,
				refreshTime);
		MatcherAssert.assertThat(returnDoc, Matchers.nullValue());
		
	}
	
	@Test
	public void verifyDocumentReturn() {
		org.mule.api.MuleEvent muleEvent = null;
		java.lang.String id = "user1";
		boolean lockDocument = false;
		int lockDuration = 0;
		boolean refreshExpirationTime = false;
		int refreshTime = 60;
		JavaMapDocument returnDoc = getConnector().getDocument(muleEvent, id, lockDocument, lockDuration, refreshExpirationTime,
				refreshTime);
		MatcherAssert.assertThat(returnDoc, Matchers.notNullValue());
		MatcherAssert.assertThat(returnDoc.getCas(), Matchers.notNullValue());
		MatcherAssert.assertThat(returnDoc.getId(), Matchers.equalTo("user1"));
		MatcherAssert.assertThat(returnDoc.getContent().get("state").toString(), Matchers.equalTo("DE"));
		
	}
	
	/**
	 * Tests the Pessimistic locking of a document. Locked document can only be updated if CAS matches with the value stored in server.
	 * Incorrect CAS value in update should result in exception.
	 */
	@Test
	public void verifyDocumentLockException() {
		org.mule.api.MuleEvent muleEvent = null;
		java.lang.String id = "user1";
		boolean lockDocument = true;
		int lockDuration = 15;
		boolean refreshExpirationTime = false;
		int refreshTime = 60;
		JavaMapDocument returnDoc = getConnector().getDocument(muleEvent, id, lockDocument, lockDuration, refreshExpirationTime,
				refreshTime);
		
		//Change the CAS and update the object. It should throw an exception.
		
		JavaMapDocument newDoc = returnDoc.clone();
		
		newDoc.setCas(1l);
		newDoc.getContent().put("city", "anything");
		
		boolean asExpected = false;
		
		try {

			getConnector().updateDocument(muleEvent, newDoc);
			
		} catch (CASMismatchException e) {
			asExpected = true;
			getConnector().unlockDocument(muleEvent, returnDoc.getId(), returnDoc.getCas());
		}

		MatcherAssert.assertThat(asExpected, Matchers.equalTo(Boolean.TRUE));
	}
	
	/**
	 * Tests the Pessimistic locking of a document. Locked document can only be updated if CAS matches with the value stored in server.
	 * Incorrect CAS value in update should result in exception.
	 */
	@Test
	public void verifyDocumentLockNoException() {
		org.mule.api.MuleEvent muleEvent = null;
		java.lang.String id = "user1";
		boolean lockDocument = true;
		int lockDuration = 15;
		boolean refreshExpirationTime = false;
		int refreshTime = 60;
		JavaMapDocument returnDoc = getConnector().getDocument(muleEvent, id, lockDocument, lockDuration, refreshExpirationTime,
				refreshTime);
		//Change an attribute, keeping CAS same.
		returnDoc.getContent().put("city", "anything");
	
		JavaMapDocument returnDoc2 = getConnector().updateDocument(muleEvent, returnDoc);
			
		MatcherAssert.assertThat(returnDoc2.getContent().get("city").toString(), Matchers.equalTo("anything"));
	}

}

src/main/java/org/mule/modules/couchbase/model/CbMapDocument.java
package org.mule.modules.couchbase.model;

import java.util.HashMap;
import java.util.Map;

import org.mule.api.annotations.Required;
import org.mule.api.annotations.display.FriendlyName;
import org.mule.api.annotations.display.Placement;
import org.mule.api.annotations.param.Default;

import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;

/**
 * This is wrapper pojo over {@link JsonDocument}. Document content is represented as Java {@link Map}.
 * @author manik
 *
 */
public class CbMapDocument implements Cloneable {
	/**
	 * Unique document identifier
	 */
	@Required
	@FriendlyName("Unique ID")
	@Placement(group="Document Metadata",order=1)
	private String id;
	
	/**
	 * Document CAS id used for concurrent locking. For server to consider this CAS, value must be >= 0.
	 */
	@FriendlyName("CAS")
	@Placement(group="Document Metadata",order=2)
	@Default("0")
	private long cas;
	
	/**
	 * Document expiration time in seconds. For server to consider expiry, value must be >= 0.
	 */
	@FriendlyName("Expiration Time")
	@Placement(group="Document Metadata",order=3)
	@Default("-1")
	private int expiry;
	
	/**
	 * {@link Map} representing the document content.
	 */
	@FriendlyName("Content Map")
	@Required
	@Placement(group="Document Content",order=1)
	private Map<String, Object> content;
	
	
	public CbMapDocument(){
		
	}
	
	public CbMapDocument(String id){
		this(id, null);
	}
	
	public CbMapDocument(String id, Map<String, Object> content){
		setId(id);
		setContent(content);
	}
	
	
	public String getId() {
		return id;
	}
	public void setId(String id) {
		this.id = id;
	}
	public long getCas() {
		return cas;
	}
	public void setCas(long cas) {
		this.cas = cas;
	}
	public int getExpiry() {
		return expiry;
	}
	public void setExpiry(int expiry) {
		this.expiry = expiry;
	}
	public Map<String, Object> getContent() {
		return content;
	}
	public void setContent(Map<String, Object> content) {
		this.content = content;
	}
	
	/**
	 * Creates an instance of {@link CbMapDocument} from {@link JsonDocument}.
	 * @param jsonDocument
	 * @return {@link CbMapDocument}
	 */
	public static CbMapDocument fromJsonDocument(JsonDocument jsonDocument){
		CbMapDocument cbMapDocument = new CbMapDocument();
		cbMapDocument.setCas(jsonDocument.cas());
		cbMapDocument.setExpiry(jsonDocument.expiry());
		cbMapDocument.setId(jsonDocument.id());
		if(jsonDocument.content() != null) {
			cbMapDocument.setContent(jsonDocument.content().toMap());
		}
		
		return cbMapDocument;
	}
	
	/**
	 * Creates {@link JsonDocument} from this {@link CbMapDocument}. 
	 * @return {@link JsonDocument}
	 */
	public JsonDocument toJsonDocument(){
		JsonObject jsonObject = JsonObject.from(this.getContent());
		
		JsonDocument document = null;
		
		if (expiry >= 0) {
			document = JsonDocument.create(this.getId(), this.getExpiry(), jsonObject, this.getCas());
		} else {
			document = JsonDocument.create(getId(), jsonObject, getCas());
		}
		
		return document;
		
	}
	
	@Override
	public String toString() {
		return "Id: " + this.id + ", CAS: "+ this.cas + ", Expiry: " + this.expiry + ", Content: " + this.content.toString();
	}
	
	@SuppressWarnings("unchecked")
	@Override
	public CbMapDocument clone() {
		CbMapDocument newDoc = new CbMapDocument();
		newDoc.setId(this.getId());
		newDoc.setCas(this.getCas());
		newDoc.setExpiry(getExpiry());
		newDoc.setContent((Map<String, Object>) ((HashMap<String, Object>)getContent()).clone());
		return newDoc;
	}

	/**
	 * Checks if Content is null or empty.
	 * @return
	 */
	public boolean hasContent(){
		return content == null || content.isEmpty();
	}
	
}


src/main/java/org/mule/modules/couchbase/model/JavaMapDocument.java
/**
 * The software in this package is published under the terms of the Apache v2.0 license,
 * a copy of which has been included with this distribution in the LICENSE.md file.
 */
package org.mule.modules.couchbase.model;

import java.util.HashMap;
import java.util.Map;

import org.mule.api.annotations.Required;
import org.mule.api.annotations.display.FriendlyName;
import org.mule.api.annotations.display.Placement;
import org.mule.api.annotations.param.Default;

import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;

/**
 * This is wrapper pojo over {@link JsonDocument}. Document content is represented as Java {@link Map}.
 * @author manik
 *
 */
public class JavaMapDocument implements Cloneable {
	/**
	 * Unique document identifier
	 */
	@Required
	@FriendlyName("Unique ID")
	@Placement(group="Document Metadata",order=1)
	private String id;
	
	/**
	 * Document CAS id used for concurrent locking. For server to consider this CAS, value must be >= 0.
	 */
	@FriendlyName("CAS")
	@Placement(group="Document Metadata",order=2)
	@Default("0")
	private long cas;
	
	/**
	 * Document expiration time in seconds. For server to consider expiry, value must be >= 0.
	 */
	@FriendlyName("Expiration Time")
	@Placement(group="Document Metadata",order=3)
	@Default("-1")
	private int expiry;
	
	/**
	 * {@link Map} representing the document content.
	 */
	@FriendlyName("Content Map")
	@Required
	@Placement(group="Document Content",order=1)
	private Map<String, Object> content;
	
	
	public JavaMapDocument(){
		
	}
	
	public JavaMapDocument(String id){
		this(id, null);
	}
	
	public JavaMapDocument(String id, Map<String, Object> content){
		setId(id);
		setContent(content);
	}
	
	
	public String getId() {
		return id;
	}
	public void setId(String id) {
		this.id = id;
	}
	public long getCas() {
		return cas;
	}
	public void setCas(long cas) {
		this.cas = cas;
	}
	public int getExpiry() {
		return expiry;
	}
	public void setExpiry(int expiry) {
		this.expiry = expiry;
	}
	public Map<String, Object> getContent() {
		return content;
	}
	public void setContent(Map<String, Object> content) {
		this.content = content;
	}
	
	/**
	 * Creates an instance of {@link JavaMapDocument} from {@link JsonDocument}.
	 * @param jsonDocument
	 * @return {@link JavaMapDocument}
	 */
	public static JavaMapDocument fromJsonDocument(JsonDocument jsonDocument){
		JavaMapDocument javaMapDocument = new JavaMapDocument();
		javaMapDocument.setCas(jsonDocument.cas());
		javaMapDocument.setExpiry(jsonDocument.expiry());
		javaMapDocument.setId(jsonDocument.id());
		if(jsonDocument.content() != null) {
			javaMapDocument.setContent(jsonDocument.content().toMap());
		}
		
		return javaMapDocument;
	}
	
	/**
	 * Creates {@link JsonDocument} from this {@link JavaMapDocument}. 
	 * @return {@link JsonDocument}
	 */
	public JsonDocument toJsonDocument(){
		JsonObject jsonObject = JsonObject.from(this.getContent());
		
		JsonDocument document = null;
		
		if (expiry >= 0) {
			document = JsonDocument.create(this.getId(), this.getExpiry(), jsonObject, this.getCas());
		} else {
			document = JsonDocument.create(getId(), jsonObject, getCas());
		}
		
		return document;
		
	}
	
	@Override
	public String toString() {
		return "Id: " + this.id + ", CAS: "+ this.cas + ", Expiry: " + this.expiry + ", Content: " + this.content.toString();
	}
	
	@SuppressWarnings("unchecked")
	@Override
	public JavaMapDocument clone() {
		JavaMapDocument newDoc = new JavaMapDocument();
		newDoc.setId(this.getId());
		newDoc.setCas(this.getCas());
		newDoc.setExpiry(getExpiry());
		newDoc.setContent((Map<String, Object>) ((HashMap<String, Object>)getContent()).clone());
		return newDoc;
	}

	/**
	 * Checks if Content is null or empty.
	 * @return
	 */
	public boolean hasContent(){
		return content == null || content.isEmpty();
	}
	
}


src/main/java/org/mule/modules/couchbase/helper/DocumentsHelper.java
package org.mule.modules.couchbase.helper;

import java.util.Map;

import org.mule.modules.couchbase.CouchbaseConnector;
import org.mule.modules.couchbase.model.CbMapDocument;

import com.couchbase.client.java.Bucket;
import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;

public class DocumentsHelper {
	
	public static Map<String, Object> getDocument(final String id, CouchbaseConnector connector){
		Bucket bucket = connector.openBucket();
		
		JsonDocument document = null;
		document = bucket.get(id);
		
		Map<String, Object> doc = document.content().toMap();
		
		return doc;
	}
	
	public static Map<String, Object> getDocumentAndLock(final String id, CouchbaseConnector connector, int lockTime){
		Bucket bucket = connector.openBucket();
		
		JsonDocument document = null;
		
		document = bucket.getAndLock(id,lockTime);
		
		Map<String, Object> doc = document.content().toMap();
		
		return doc;
	}
	
	public static Map<String, Object> getDocumentAndTouch(final String id, CouchbaseConnector connector, int touchDuration){
		Bucket bucket = connector.openBucket();
		
		JsonDocument document = null;
		
		document = bucket.getAndTouch(id, touchDuration);
		
		Map<String, Object> doc = document.content().toMap();
		
		return doc;
	}
	
	
	public static CbMapDocument upsertDocument(final String id, final Map<String, Object> contentMap, CouchbaseConnector connector){
		Bucket bucket = connector.openBucket();
		
		JsonObject jsonObject = JsonObject.from(contentMap);
		
		JsonDocument document = JsonDocument.create(id, jsonObject);
		
		document = bucket.upsert(document);
		
		return CbMapDocument.fromJsonDocument(document);
		
	}
}


src/main/java/org/mule/modules/couchbase/CouchbaseConnector.java
/**
 * The software in this package is published under the terms of the Apache v2.0 license,
 * a copy of which has been included with this distribution in the LICENSE.md file.
 */
package org.mule.modules.couchbase;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;

import org.mule.api.MuleContext;
import org.mule.api.MuleEvent;
import org.mule.api.annotations.Config;
import org.mule.api.annotations.Connector;
import org.mule.api.annotations.Processor;
import org.mule.api.annotations.display.FriendlyName;
import org.mule.api.annotations.display.Placement;
import org.mule.api.annotations.display.Summary;
import org.mule.api.annotations.param.Default;
import org.mule.extension.annotations.param.Optional;
import org.mule.modules.couchbase.config.CouchbaseConnectorConfig;
import org.mule.modules.couchbase.model.JavaMapDocument;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.messaging.MessagingException;

import com.couchbase.client.java.Bucket;
import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonArray;
import com.couchbase.client.java.document.json.JsonObject;
import com.couchbase.client.java.error.DocumentDoesNotExistException;
import com.couchbase.client.java.query.N1qlQuery;
import com.couchbase.client.java.query.N1qlQueryResult;
import com.couchbase.client.java.query.N1qlQueryRow;

@Connector(name="couchbasedb", friendlyName="Couchbase DB",
			minMuleVersion="3.5.3",keywords="Couchbase, Database",
			description="Mule Connector for Couchbase NoSQL database")
public class CouchbaseConnector{
	
	private Logger LOG = LoggerFactory.getLogger(CouchbaseConnector.class); 

    @Config
    CouchbaseConnectorConfig config;
    
    @Inject
    private MuleContext muleContext;
    
	public MuleContext getMuleContext() {
		return muleContext;
	}

	public void setMuleContext(MuleContext muleContext) {
		this.muleContext = muleContext;
	}

	public CouchbaseConnectorConfig getConfig() {
        return config;
    }

    public void setConfig(CouchbaseConnectorConfig config) {
        this.config = config;
    }
    
    public Bucket openBucket(){
    	return config.openBucket();
    }
    
    /**
     * This method retrieves document with given id from couchbase database. It can optionally lock and/or refresh the expiration time of document.
     * @param muleEvent
     * @param id Unique identifier of document
     * @param lockDocument
     * @param lockDuration
     * @param refreshExpirationTime
     * @param refreshTime
     * @return {@link JavaMapDocument} containing document id, cas, expiry and content from database.
     * @see Bucket#get(com.couchbase.client.java.document.Document)
     * @see Bucket#getAndLock(com.couchbase.client.java.document.Document, int)
     * @see Bucket#getAndTouch(com.couchbase.client.java.document.Document)
     */
    @Processor(friendlyName="Get Document")
    public JavaMapDocument getDocument(MuleEvent muleEvent, String id, 
    		@Placement(group="Document Locking (pessimistic)") @Default("false") boolean lockDocument,
    		@Placement(group="Document Locking (pessimistic)") @Max(30) @Min(0) @Summary("Define seconds for which document should be write-locked. This creates pissimistic locking. Read more on couchbase documentation.") @Default("15") int lockDuration,
    		@Placement(group="Refresh Expiration Time") @Default("false") boolean refreshExpirationTime,
    		@Placement(group="Refresh Expiration Time") @Summary("Time in seconds. If you specify an expiration time greater than 30 days in seconds (60 seconds * 60 minutes * 24 hours * 30 days = 2,592,000 seconds) it is considered an absolute time stamp instead of a relative one") 
    					@Default("0") int refreshTime){
   		
   		Bucket bucket = openBucket();
		
		JsonDocument document = null;
		
		if(lockDocument){
			if(refreshExpirationTime) bucket.touch(id, refreshTime);
   			document = bucket.getAndLock(id, lockDuration);
   		} else if (refreshExpirationTime) {
   			document = bucket.getAndTouch(id, refreshTime);
   		} else {
   			document =  bucket.get(id);
   		}
		
		if(document == null){
			return null;
		} else {
			JavaMapDocument doc = JavaMapDocument.fromJsonDocument(document);
			return doc;
		}
		
    }
   	
   	/**
   	 * This method update/inserts the JSON document for given id into couchbase database.
   	 * @param muleEvent
   	 * @param javaMapDocument {@link JavaMapDocument} containing document id and content to be upserted in database.
   	 * @return {@link JavaMapDocument} containing document id, cas, expiry and content from database.
   	 * @see com.couchbase.client.java.Bucket#upsert(com.couchbase.client.java.document.Document)
   	 */
   	@Processor(friendlyName="UpSert Document")
   	public JavaMapDocument upsertDocument(MuleEvent muleEvent, @Placement(group="Document to upsert") @FriendlyName("Document") @Summary("Specify the document to be upserted.") JavaMapDocument javaMapDocument){
   		
   		Bucket bucket = openBucket();
		
		JsonDocument document = javaMapDocument.toJsonDocument();
		
		document = bucket.upsert(document);
		
		JavaMapDocument returnDocument = JavaMapDocument.fromJsonDocument(document);
   		
   		return returnDocument;
   	}
   	
   	/**
   	 * This method tries to insert the JSON document for given id into couchbase database. It will throw an exception if document with given id already exists in database.
   	 * @param muleEvent
   	 * @param javaMapDocument {@link JavaMapDocument} containing document id and content to be inserted in database.
   	 * @return {@link JavaMapDocument} containing document id, cas, expiry and content from database.
   	 * @see com.couchbase.client.java.Bucket#insert(com.couchbase.client.java.document.Document)
   	 */
   	@Processor(friendlyName="Insert Document")
   	public JavaMapDocument insertDocument(MuleEvent muleEvent, @Placement(group="Document to Insert") @FriendlyName("Document") @Summary("Specify the document to be inserted.") JavaMapDocument javaMapDocument){
   		
   		Bucket bucket = openBucket();
		
		JsonDocument document = javaMapDocument.toJsonDocument();
		
		document = bucket.insert(document);
		
		JavaMapDocument returnDocument = JavaMapDocument.fromJsonDocument(document);
   		
   		return returnDocument;
   		
   	}
   	
   	/**
   	 * This method tries to update an existing JSON document for given id into couchbase database. It will throw an exception if document with given id Does not exist.
   	 * 
   	 * @param muleEvent
   	 * @param javaMapDocument {@link JavaMapDocument} containing document id and content to be updated in database.
   	 * @return {@link JavaMapDocument} containing document id, cas, expiry and content from database.
   	 * @see com.couchbase.client.java.Bucket#replace(com.couchbase.client.java.document.Document)
   	 */
   	@Processor(friendlyName="Update Document")
   	public JavaMapDocument updateDocument(MuleEvent muleEvent, @Placement(group="Document to Update") @FriendlyName("Document") @Summary("Specify the document to be updated.") JavaMapDocument javaMapDocument){
   		
   		Bucket bucket = openBucket();
		
		JsonDocument document = javaMapDocument.toJsonDocument();
		
		if(!bucket.exists(document.id())) {
			throw new DocumentDoesNotExistException("Document with id "+ document.id() +" does not exist.");
		}
		
		document = bucket.replace(document);
		
		JavaMapDocument returnDocument = JavaMapDocument.fromJsonDocument(document);
   		
   		return returnDocument;
   		
   	}
   	
   	/**
   	 * This method removes the document with given id from database. Returned {@link JavaMapDocument} only contains ID and CAS value set as document is already removed from server.
   	 * 
   	 * @param muleEvent
   	 * @param javaMapDocument {@link JavaMapDocument} containing id of the document to be removed from server. Content value is ignored.
   	 * @return {@link JavaMapDocument} containing document id, cas, expiry and content from database.
   	 * @see com.couchbase.client.java.Bucket#remove(com.couchbase.client.java.document.Document)
   	 */
   	@Processor(friendlyName="Delete Document")
   	public boolean deleteDocument(MuleEvent muleEvent, @Placement(group="Document to Delete") @FriendlyName("Document") @Summary("Specify the document to be deleted.") JavaMapDocument javaMapDocument){
   		
   		Bucket bucket = openBucket();
		
   		if(!bucket.exists(javaMapDocument.getId())) {
			LOG.error("Document with id "+ javaMapDocument.getId() +" does not exist.", new DocumentDoesNotExistException("Document with id "+ javaMapDocument.getId() +" does not exist."));
			return false;
		}
   		
		JsonDocument document = bucket.remove(javaMapDocument.getId());
		
		try {
			JavaMapDocument.fromJsonDocument(document);
		} catch (DocumentDoesNotExistException notExists) {
			LOG.info("Couchbase Delete: Document to delete does not exist: "+ javaMapDocument.getId());
			return false;
		}catch (Exception otherEx) {
			LOG.info("Couchbase Delete: Unable to delete document: "+ javaMapDocument.getId());
			return false;
		}
		
   		
   		return true;
   		
   	}
   	
   	/**
   	 * This method unlocks the document that was previously locked (pessimistic). Document can be unlocked by providing either document ID and CAS OR whole document. Document can be retrieved with get document method. 
   	 * 
   	 * @param muleEvent
   	 * @param id Unique Id of the document
   	 * @param cas Long CAS value that must match with the one stored in database.
   	 * @return {@link Boolean} if document is successfully unlocked
   	 * @see com.couchbase.client.java.Bucket#unlock(com.couchbase.client.java.document.Document)
   	 */
   	@Processor(friendlyName="Unlock Document")
   	public boolean unlockDocument(MuleEvent muleEvent, @Placement(group="Document to be Unlocked") @FriendlyName("Document ID") @Summary("Specify the document Id to be unlocked.") String id,
   			@Placement(group="Document to be Unlocked") @FriendlyName("Document CAS") @Summary("Document CAS must match with the one stored in database.") Long cas){
   		
   		Bucket bucket = openBucket();
   		
   		boolean unlocked = bucket.unlock(id, cas);
		
   		return unlocked;
   		
   	}
   	
   	/**
   	 * Runs the N1QL query and result the result set. Optionally, either positional or named parameters can be provided.
   	 * @param muleEvent
   	 * @param query to be executed.
   	 * @param positionalParams {@link List} containing positional parameters to replace place holders in query. Takes precedence over named parameters.
   	 * @param namedParams {@link Map} containing named parameters to be replaced in query.
   	 * @return {@link List} containing the documents in map.
   	 */
   	@Processor(friendlyName="Execute Query")
   	public List<Map<String, Object>> executeQuery(MuleEvent muleEvent, String query, @Optional @FriendlyName("Positional Parameters") 
   				@Summary("For positional parameters, specify list of param values. For named parameters, specify Map of key-value pairs.") Object params){
   		
   		Bucket bucket = openBucket();
   		
   		bucket.bucketManager().createN1qlPrimaryIndex(true, false);
   		
   		N1qlQueryResult result;
   		
   		if(params != null && params instanceof List){
   			result = bucket.query(N1qlQuery.parameterized(query, JsonArray.from(params)));
   		} else if (params != null && params instanceof Map) {
   			result = bucket.query(N1qlQuery.parameterized(query, JsonObject.from((Map)params)));
   		} else {
   			result = bucket.query(N1qlQuery.simple(query));
   		}
   		
   		if(!result.parseSuccess()){
   			LOG.error("Failed to parse the query with errors: " + result.errors());
   			throw new MessagingException("Failed to parse the N1QL Query: "+ query);
   		}
   		
   		if(!result.finalSuccess()){
   			LOG.error("Failed to execute the query with errors: "+ result.errors());
   			throw new MessagingException("Failed to execute the N1QL Query: "+ query);
   		}
   		
   		List<Map<String, Object>> resultSet = new ArrayList<Map<String,Object>>();
   		
   		for (N1qlQueryRow row : result.allRows()){
   			resultSet.add(row.value().toMap());
   		}
   		
   		return resultSet;
   		
   	}


}